UI Screenshots script

DingoEatingFuzz opened the pull request:
This is a puppeteer-based node script for automatically capturing UI screenshots to use in docs and guides and whatnot.

Since puppeteer depends on headless chrome which can be it's hassle to setup, I have provided a docker-based way to run the script via `make ui-screenshots` in the website dir.

But since that docker container is based on Ubuntu and Ubuntu has an objectively worse system typeface than macOS, I also made sure to add `make ui-screenshots-local` which will run the script locally.

The next step is to rewrite some UI guides since so much as changed since they were authored.

johncowen commented on ui/app/components/job-page/parts/recent-allocations.js:
```diff
@@ -3,6 +3,8 @@ import { computed } from '@ember/object';
 import PromiseArray from 'nomad-ui/utils/classes/promise-array';
 
 export default Component.extend({
+  classNames: ['boxed-section'],
```
Guessing this might have made it in here by mistake?

johncowen replied:
Ah ignore this, I've just seen where you are using `.boxed-section` further down? Actually if this class is just for testing is there a more appropriate way of doing it? Maybe not, but thought I'd ask. I'm guessing you are running this as as a production build and therefore can't use the `data-test-* attributes?

DingoEatingFuzz replied:
Nah, this class was added to fix a bug where two elements on the parameterized-child job page had no margin between them.

That it's used in the puppeteer script is semi-coincidence and semi-sadness. As you suspected, I'm not using the `data-test-*` attributes because they aren't available. Not because it's a production build but because `data-test-*` is only in test builds 😢 It might be worth configuring that to show up in dev builds.

johncowen replied:
Oh weird yeah, 'pretty sure' they show up in our dev builds, though I might be misremembering, I probably have a slightly different config to nomad though. What would be ace is if you could use PageObjects here. 

This is one of those classic things that changes without people realising though, if it could share the PageObjects/`data-test-*` things that would be great.

Just bouncing this around for the chat more than anything though, I agree with your goal with this 🔨

DingoEatingFuzz replied:
Using the same PageObjects would be super awesome. Granted the ember bindings would have to be replaced with puppeteer ones...and somehow shared across the UI code base and this script...but I can see it happening and being much better than what I'm doing right now 🙈

johncowen commented on website/scripts/screenshots/src/index.js:
```diff
@@ -0,0 +1,180 @@
+
+  // Give Mirage a chance to settle
+  console.log("Waiting for Mirage...");
+  await wait(5000);
```
I'm sure this is probably fine, but just wondering if there anyway you can check for when Mirage is ready?

DingoEatingFuzz replied:
I think there is some `waitForElement`-like features in puppeteer? My goal was to spike something out quickly rather than well 😄 This is by no means the finest of engineering, but since it's just a script for internal use I figured this sledgehammer approach was adequate.

johncowen replied:
Yep perfect, gotcha 👍

johncowen commented on ui/config/environment.js:
```diff
@@ -23,7 +23,7 @@ module.exports = function(environment) {
       blockingQueries: true,
       mirageScenario: 'smallCluster',
-      mirageWithNamespaces: false,
+      mirageWithNamespaces: true,
```
Just wondering what this does?

DingoEatingFuzz replied:
It creates fixtures for [Nomad namespaces](https://www.nomadproject.io/guides/governance-and-policy/namespaces.html) which manifest as a dropdown in the left-hand gutter menu.

johncowen commented on ui/mirage/scenarios/default.js:
```diff
@@ -11,6 +11,7 @@ const allScenarios = {
   massiveCluster,
   allJobTypes,
   allNodeTypes,
+  allTheThings,
```
Although I love the name here, would it be better to call this something to do with screengrabs? Maybe like 'screengrabScenario'? I'm assuming this scenario is especially for taking the screengrabs? Maybe just a comment if you want to keep the name and potentially use it for other things also?

DingoEatingFuzz replied:
I definitely made this one for the screengrabs, and the puppeteer script will break if this one isn't used, but I also think this scenario is valid beyond screengrabs.

I agree that the coupling needs to be communicated in some manner.

DingoEatingFuzz commented:
Ooookay. I changed the `allTheThings` scenario to `everyFeature` and added a warning in the puppeteer script to help with debugging in the event it's not already set and the script goes 💥 

@backspace and I chatted about it and we both agree that having some sort of envvar here that could be set in the make target instead would be better, but I'm resisting the temptation to do that now for the sake of moving on with revising the UI guide this whole thing spun out of.

![image](https://user-images.githubusercontent.com/174740/59458173-ccc8c580-8dce-11e9-9508-1963ae53c4f1.png)

johncowen commented:
Yup yup, sounds good! Sort it later though, don't want to hold you up

John
